### ISSUE TITLE
Refactor the Audience Tiles to use new pivot report infrastructure

### ISSUE THREAD

**Issue by benbowler** [google/site-kit-wp#8726]
## Feature Description

As discussed on [Slack](https://10up.slack.com/archives/CBKKQEBR9/p1712306329670989), while we are initially implementing the Audience Tiles to use separate, per-audience reports to retrieve the cities and "top content" metrics (one report per audience per metric), we can reduce the number of reports needed by using [pivot reports](https://developers.google.com/analytics/devguides/reporting/data/v1/rest/v1beta/properties/runPivotReport). This will allow us to retrieve metric data for all of the audiences in a single report (i.e., one report per metric).

We should add support for running pivot reports, and refactor the Audience Tiles to use them.

This is not critical to the release and can be done post-launch, hence the P2 priority.

Please also note this comment relating to the `AudienceTile` and `AudienceTiles` refactoring: https://github.com/google/site-kit-wp/issues/8484#issuecomment-2059400367

**This ticket has been split from #8484, and only concerns updating the Audience Tiles to use the new pivot report infrastructure, which will be completed in that ticket.**

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation Brief

* [ ] 

### Test Coverage

* 

## QA Brief

* 

## Changelog entry

*

**Comment by benbowler** [comment]
FYI, this ticket was split from #8484 on 17 May 2024. #8484 is a blocker.

**Comment by techanvil** [comment]
Hey @benbowler, thanks for drafting this IB. A few points:

- The first report in `AudienceTiles` doesn't seem to need the update - it's already being retrieved via `getReport()` rather than `getReportForAllAudiences()`, and seems fine as it is AFAICT.
- The remaining reports should be updated to use the `getPivotReport()` selector in line with the proposed changes to https://github.com/google/site-kit-wp/issues/8484, assuming they have been made.
- The report with options `topContentPageTitlesReportOptions` should also be covered in the IB (including the refactoring of the restructuring code).
- A minor point but the report option code blocks are slightly misaligned in places.
- Another minor one, I think the point in **Data Mock Updates** about updating "`AudienceTile` stories to use this new data-mock functionality" should refer to "the new data-mock functionality introduced in #8484" for the sake of clarity.

**Comment by techanvil** [comment]
Thanks for the update @benbowler. It looks like these two points have been missed, though, please can you take a look?

> - The report with options `topContentPageTitlesReportOptions` should also be covered in the IB (including the refactoring of the restructuring code).
> - A minor point but the report option code blocks are slightly misaligned in places.

**Comment by ankitrox** [comment]
This is ticket is ready to be moved to CR, but it should be done once #8484 is merged and tested successfully as #8484 is the dependency for this and any changes in 8484 may lead to the changes in this task.

**Comment by techanvil** [comment]
Thanks @ankitrox - in cases like this, we don't tend to wait for the dependency to go through QA, simply having it merged is enough to move this to CR. Please remember to merge `develop` into this issue's branch, and retarget the PR first :)

**Comment by ankitrox** [comment]
Thanks @techanvil . I will move this to CR as soon as #8484 gets merged. I will also take care to update base branch and merge it to feature one.

Thank you.

**Comment by techanvil** [comment]
Hi @ankitrox, I have left a few comments on the PR, please take a look.

Additionally, reviewing the QAB I'd note that the code snippets aren't really needed - the feature is well enough integrated that it should be completely testable via the UI.

So, I'd suggest a bit of a rewrite with that in mind. Essentially it just needs a bit of a smoke test of the audience tiles to ensure that nothing has changed.

It would also be a useful to provide a brief QA:Eng section to verify that pivot reports can be seen in the devtools network tab.

**Comment by wpdarren** [comment]
@kelvinballoo as per our chat on Slack. I have assigned this to you. Once you have completed your testing please unassign yourself and we'll get an engineer to look over it for the `QA:Eng` part.

**Comment by kelvinballoo** [comment]
### QA Update ⚠️ 

Despite being simple, this is a bit difficult to smoke test because a lot of things could be because of other tickets not deployed yet. 

The tiles are appearing as how they were previously. ✅ 
That said, the tricky part is about verifying whether data is intact .
Here are some observations to be reviewed.
(For context, I used oi.ie property for testing)

**ITEM 1:**
`A round of smoke test to ensure that this feature has not changed anything in the Audience tiles widget. Also that data should be intact within those tiles.`
- What are we actually meaning by intact data?
- Verifying the data from Analytics doesn't seem to correspond. Refer to the video attached.
<Details>

https://github.com/google/site-kit-wp/assets/125576926/f9d8d6aa-59ba-4156-bc86-9a022e731367


</Details>
__________________________________________________________________

**ITEM 2:**
`Ensure that top cities are displaying in the widget. Verify the data from Google Analytics dashboard that it is displaying correctly.`
The top cities are displaying but it's showing one of them as '**(not set)**'

<Details>
The second city is (not set)
<img width="1174" alt="Screenshot 2024-07-03 at 18 23 44" src="https://github.com/google/site-kit-wp/assets/125576926/247da0b6-c637-4eb9-90df-ba0582c70545">

On analytics, I could see the cities. Is it something to be set up in the BE?
<img width="366" alt="Screenshot 2024-07-03 at 18 25 07" src="https://github.com/google/site-kit-wp/assets/125576926/4497151e-f616-45a2-a407-3b81095be6e2">

</Details>

__________________________________________________________________

**ITEM 3:**

`Ensure that "Top content by pageviews" are displaying in the widget. Verify the data from Google Analytics dashboard that it is displaying correctly.`
I've verified the number from GA and the tile for the 'All visitors' at the 'Top content by pageviews' section and they matched for the first 2 but not the third. 

<Details>
The 3rd item is 245 on the tile.
<img width="1157" alt="Screenshot 2024-07-03 at 18 33 54" src="https://github.com/google/site-kit-wp/assets/125576926/c073ba91-36dd-4cfd-a09c-15a464c0c03e">

The 3rd item is 403 in Analytics
<img width="1005" alt="Screenshot 2024-07-03 at 18 34 05" src="https://github.com/google/site-kit-wp/assets/125576926/2bc3db26-27da-4e69-be16-2c4d2239787f">

</Details>

__________________________________________________________________

**ITEM 4:**
I feel the results of the tiles are not properly matched.

Refer to the attached video for details
<Details>

https://github.com/google/site-kit-wp/assets/125576926/e0a9b2d6-95a0-4813-8b8f-074075a04258


</Details>

**Comment by techanvil** [comment]
Hey @kelvinballoo, what I'd suggest for the smoke test is to refer back to this [comment](https://github.com/google/site-kit-wp/issues/8763#issuecomment-2147935477) I left on 8763 where I point out that we can see the same metrics elsewhere 

> It's worth noting that all of the metrics for an audience are viewable elsewhere in Site Kit, so when viewing the "all visitors" audience, we can cross reference other widgets as a sanity check for the figures.
> To illustrate the point, here's a map of the metrics on the Audience Tile to the same metrics on Key Metric tiles and the All Traffic widget:
> 
> ![image](https://github.com/google/site-kit-wp/assets/18395600/2e7b7d41-1ed0-454c-86c6-2c12d863f0ef)

You can cross reference the metrics like this to check if anything has changed.

If the wider dashboard matches the All Visitors audience tile but you find discrepancies between the figures and what you see on Analytics, this can be raised as a separate issue for investigation.

**Comment by kelvinballoo** [comment]
### QA Update ✅ 

Thanks @techanvil , that really helps.

I've verified this way and I can confirm that the tiles data is matching the data on the other part of the SK dashboard.
Marking this as Pass but not moving this to approval yet since it's a 'QA Eng' ticket.

<Details>
<img width="1184" alt="Screenshot 2024-07-04 at 14 05 15" src="https://github.com/google/site-kit-wp/assets/125576926/c2f59044-5abc-43a6-9512-1d2b0da2a180">
<img width="1183" alt="Screenshot 2024-07-04 at 14 05 23" src="https://github.com/google/site-kit-wp/assets/125576926/82d91e02-9f7c-4204-a2bf-8727ea26787a">

</Details>

**Comment by techanvil** [comment]
## QA:Eng :x:

While performing QA:Eng for this issue, I've had a horrible realisation that pivot reports aren't actually suitable for the audience tiles in the way we thought they were.

This has come to light while examining the report output for audiences which have visitors from different cities.

As can be seen in the current audience data for `oi.ie`, the "Mauritius users" audience has visitors from London:
![image](https://github.com/google/site-kit-wp/assets/18395600/da292d9a-5989-474f-bb53-91a5ffdb6769)

While "Returning visitors" has visitors from Dublin, "(not set)", and Cork:
![image](https://github.com/google/site-kit-wp/assets/18395600/88045088-187c-4dcc-adbf-3bda4f028f26)

However when _both_ of these audiences are visible, we only see Dublin, "(not set)", and Cork for both audiences, with no sign of London:

![image](https://github.com/google/site-kit-wp/assets/18395600/7eac6c75-1bc7-4c24-810e-2b164e4e81a5)

This was admittedly made harder to spot by the fact we have the open "2nd tile has no data" bug, https://github.com/google/site-kit-wp/issues/8921 which adds a bit of ambiguity. However the drilling into the actual reports as returned by the API indicates this is indeed the case and the reports are simply not returning the data in the way that we would like.

Essentially, each pivot in the report effectively results in a sub-query to retrieve all the values for that query similar to a regular report where the `dimensionFilter` and so forth come into play. It's then using the returned values as the values to pivot on. So when retrieving the top 3 cities while both audiences are included in the filter, it's always returning Dublin, "(not set)", and Cork, as these cities all have more `totalUsers` than London.

It's only when we remove "Returning visitors" from the query - meaning our report is filtering results for just the "Mauritius users" audience - that London is returned, because the other cities are being excluded - they are only associated with the other audience which is no longer included in the filter.

To help to understand this further here are some illustrative reports and their responses to compare:

<details>
<summary>Report request with a filter for the "Mauritius users" audience</summary>

```json
{
  "dateRanges": [
    {
      "endDate": "2024-07-04",
      "startDate": "2024-06-07"
    }
  ],
  "dimensionFilter": {
    "andGroup": {
      "expressions": [
        {
          "filter": {
            "fieldName": "hostName",
            "inListFilter": {
              "values": ["oi.ie", "www.oi.ie"]
            }
          }
        },
        {
          "filter": {
            "fieldName": "audienceResourceName",
            "inListFilter": {
              "values": ["properties/285794360/audiences/8278021604"]
            }
          }
        }
      ]
    }
  },
  "dimensions": [
    {
      "name": "city"
    },
    {
      "name": "audienceResourceName"
    },
    {
      "name": "hostName"
    }
  ],
  "keepEmptyRows": true,
  "metrics": [
    {
      "name": "totalUsers"
    }
  ],
  "pivots": [
    {
      "fieldNames": ["city"],
      "limit": "3",
      "orderBys": [
        {
          "desc": "true",
          "metric": {
            "metricName": "totalUsers"
          }
        }
      ]
    },
    {
      "fieldNames": ["audienceResourceName"],
      "limit": "2"
    }
  ],
  "property": "properties/285794360"
}
```

</details>

<details>
<summary>Response for report filtered by the "Mauritius users" audience</summary>

```json
{
  "pivotHeaders": [
    {
      "pivotDimensionHeaders": [
        {
          "dimensionValues": [
            {
              "value": "London"
            }
          ]
        }
      ],
      "rowCount": 1
    },
    {
      "pivotDimensionHeaders": [
        {
          "dimensionValues": [
            {
              "value": "properties/285794360/audiences/8278021604"
            }
          ]
        }
      ],
      "rowCount": 1
    }
  ],
  "dimensionHeaders": [
    {
      "name": "city"
    },
    {
      "name": "audienceResourceName"
    }
  ],
  "metricHeaders": [
    {
      "name": "totalUsers",
      "type": "TYPE_INTEGER"
    }
  ],
  "rows": [
    {
      "dimensionValues": [
        {
          "value": "London"
        },
        {
          "value": "properties/285794360/audiences/8278021604"
        }
      ],
      "metricValues": [
        {
          "value": "1"
        }
      ]
    }
  ],
  "metadata": {
    "currencyCode": "USD",
    "timeZone": "America/Los_Angeles"
  },
  "kind": "analyticsData#runPivotReport"
}
```

</details>

<details>
<summary>Report request with a filter for the "Returning visitors" audience</summary>

```json
{
  "dateRanges": [
    {
      "endDate": "2024-07-04",
      "startDate": "2024-06-07"
    }
  ],
  "dimensionFilter": {
    "andGroup": {
      "expressions": [
        {
          "filter": {
            "fieldName": "hostName",
            "inListFilter": {
              "values": ["oi.ie", "www.oi.ie"]
            }
          }
        },
        {
          "filter": {
            "fieldName": "audienceResourceName",
            "inListFilter": {
              "values": ["properties/285794360/audiences/8281378703"]
            }
          }
        }
      ]
    }
  },
  "dimensions": [
    {
      "name": "city"
    },
    {
      "name": "audienceResourceName"
    },
    {
      "name": "hostName"
    }
  ],
  "keepEmptyRows": true,
  "metrics": [
    {
      "name": "totalUsers"
    }
  ],
  "pivots": [
    {
      "fieldNames": ["city"],
      "limit": "3",
      "orderBys": [
        {
          "desc": "true",
          "metric": {
            "metricName": "totalUsers"
          }
        }
      ]
    },
    {
      "fieldNames": ["audienceResourceName"],
      "limit": "2"
    }
  ],
  "property": "properties/285794360"
}
```

</details>

<details>
<summary>Response for report filtered by the "Returning visitors" audience</summary>

```json
{
  "pivotHeaders": [
    {
      "pivotDimensionHeaders": [
        {
          "dimensionValues": [
            {
              "value": "Dublin"
            }
          ]
        },
        {
          "dimensionValues": [
            {
              "value": "(not set)"
            }
          ]
        },
        {
          "dimensionValues": [
            {
              "value": "Cork"
            }
          ]
        }
      ],
      "rowCount": 151
    },
    {
      "pivotDimensionHeaders": [
        {
          "dimensionValues": [
            {
              "value": "properties/285794360/audiences/8281378703"
            }
          ]
        }
      ],
      "rowCount": 1
    }
  ],
  "dimensionHeaders": [
    {
      "name": "city"
    },
    {
      "name": "audienceResourceName"
    }
  ],
  "metricHeaders": [
    {
      "name": "totalUsers",
      "type": "TYPE_INTEGER"
    }
  ],
  "rows": [
    {
      "dimensionValues": [
        {
          "value": "Dublin"
        },
        {
          "value": "properties/285794360/audiences/8281378703"
        }
      ],
      "metricValues": [
        {
          "value": "139"
        }
      ]
    },
    {
      "dimensionValues": [
        {
          "value": "(not set)"
        },
        {
          "value": "properties/285794360/audiences/8281378703"
        }
      ],
      "metricValues": [
        {
          "value": "59"
        }
      ]
    },
    {
      "dimensionValues": [
        {
          "value": "Cork"
        },
        {
          "value": "properties/285794360/audiences/8281378703"
        }
      ],
      "metricValues": [
        {
          "value": "32"
        }
      ]
    }
  ],
  "metadata": {
    "currencyCode": "USD",
    "timeZone": "America/Los_Angeles"
  },
  "kind": "analyticsData#runPivotReport"
}
```

</details>

<details>
<summary>Report request with a filter for both audiences</summary>

```json
{
  "dateRanges": [
    {
      "endDate": "2024-07-04",
      "startDate": "2024-06-07"
    }
  ],
  "dimensionFilter": {
    "andGroup": {
      "expressions": [
        {
          "filter": {
            "fieldName": "hostName",
            "inListFilter": {
              "values": ["oi.ie", "www.oi.ie"]
            }
          }
        },
        {
          "filter": {
            "fieldName": "audienceResourceName",
            "inListFilter": {
              "values": [
                "properties/285794360/audiences/8278021604",
                "properties/285794360/audiences/8281378703"
              ]
            }
          }
        }
      ]
    }
  },
  "dimensions": [
    {
      "name": "city"
    },
    {
      "name": "audienceResourceName"
    },
    {
      "name": "hostName"
    }
  ],
  "keepEmptyRows": true,
  "metrics": [
    {
      "name": "totalUsers"
    }
  ],
  "pivots": [
    {
      "fieldNames": ["city"],
      "limit": "3",
      "orderBys": [
        {
          "desc": "true",
          "metric": {
            "metricName": "totalUsers"
          }
        }
      ]
    },
    {
      "fieldNames": ["audienceResourceName"],
      "limit": "2"
    }
  ],
  "property": "properties/285794360"
}
```

</details>

<details>
<summary>Response for report filtered by both audiences</summary>

```json
{
  "pivotHeaders": [
    {
      "pivotDimensionHeaders": [
        {
          "dimensionValues": [
            {
              "value": "Dublin"
            }
          ]
        },
        {
          "dimensionValues": [
            {
              "value": "(not set)"
            }
          ]
        },
        {
          "dimensionValues": [
            {
              "value": "Cork"
            }
          ]
        }
      ],
      "rowCount": 151
    },
    {
      "pivotDimensionHeaders": [
        {
          "dimensionValues": [
            {
              "value": "properties/285794360/audiences/8281378703"
            }
          ]
        },
        {
          "dimensionValues": [
            {
              "value": "properties/285794360/audiences/8278021604"
            }
          ]
        }
      ],
      "rowCount": 2
    }
  ],
  "dimensionHeaders": [
    {
      "name": "city"
    },
    {
      "name": "audienceResourceName"
    }
  ],
  "metricHeaders": [
    {
      "name": "totalUsers",
      "type": "TYPE_INTEGER"
    }
  ],
  "rows": [
    {
      "dimensionValues": [
        {
          "value": "Dublin"
        },
        {
          "value": "properties/285794360/audiences/8281378703"
        }
      ],
      "metricValues": [
        {
          "value": "139"
        }
      ]
    },
    {
      "dimensionValues": [
        {
          "value": "(not set)"
        },
        {
          "value": "properties/285794360/audiences/8281378703"
        }
      ],
      "metricValues": [
        {
          "value": "59"
        }
      ]
    },
    {
      "dimensionValues": [
        {
          "value": "Cork"
        },
        {
          "value": "properties/285794360/audiences/8281378703"
        }
      ],
      "metricValues": [
        {
          "value": "32"
        }
      ]
    },
    {
      "dimensionValues": [
        {
          "value": "(not set)"
        },
        {
          "value": "properties/285794360/audiences/8278021604"
        }
      ],
      "metricValues": [
        {
          "value": "0"
        }
      ]
    },
    {
      "dimensionValues": [
        {
          "value": "Cork"
        },
        {
          "value": "properties/285794360/audiences/8278021604"
        }
      ],
      "metricValues": [
        {
          "value": "0"
        }
      ]
    },
    {
      "dimensionValues": [
        {
          "value": "Dublin"
        },
        {
          "value": "properties/285794360/audiences/8278021604"
        }
      ],
      "metricValues": [
        {
          "value": "0"
        }
      ]
    }
  ],
  "metadata": {
    "currencyCode": "USD",
    "timeZone": "America/Los_Angeles"
  },
  "kind": "analyticsData#runPivotReport"
}
```

</details>

As can be seen, when just "Mauritius users" is included in the filter, the cities pivot values are as follows:

```json
    {
      "pivotDimensionHeaders": [
        {
          "dimensionValues": [
            {
              "value": "London"
            }
          ]
        }
      ],
      "rowCount": 1
    },
```

When just "Returning visitors" is included in the filter, the values are as follows:

```json
    {
      "pivotDimensionHeaders": [
        {
          "dimensionValues": [
            {
              "value": "Dublin"
            }
          ]
        },
        {
          "dimensionValues": [
            {
              "value": "(not set)"
            }
          ]
        },
        {
          "dimensionValues": [
            {
              "value": "Cork"
            }
          ]
        }
      ],
      "rowCount": 151
    },
```

When _both_ audiences are included in the filter, the values are again as follows, because they are the top 3 cities that are associated with _either_ of the audiences.

```json
    {
      "pivotDimensionHeaders": [
        {
          "dimensionValues": [
            {
              "value": "Dublin"
            }
          ]
        },
        {
          "dimensionValues": [
            {
              "value": "(not set)"
            }
          ]
        },
        {
          "dimensionValues": [
            {
              "value": "Cork"
            }
          ]
        }
      ],
      "rowCount": 151
    },
```

Note that event if we completely remove `audienceResourceName` from `dimensionFilter`, the same three cities are returned - these are the top cities across all audiences.

The "cities" case has been illustrated but the general principle applies to the "top content" metric as well.

From what I've seen so far it doesn't seem possible to use a pivot report in the way we'd like to, returning a different selection of pivot values for each audience.

It could the case that I've missed something and there is some way to achieve this, and we can investigate this further at a later date, but for now we need to revert the changes in this issue.

**Comment by techanvil** [comment]
Note that due to the above, this issue will be closed as "not planned" and not included in the forthcoming release.

**Comment by nfmohit** [comment]
The PR #8969 that reverts all changes made by this issue has been merged. As discussed above, I'm closing this issue as not planned.
